8090123: Items are no longer visible when collection is changed

karthikpandelu opened the pull request:
When a large number of items were scrolled in the `ChoiceBox`, the scrolled offset was carried forward when the list is replaced with small number of items. Hence the scroll up arrow was displayed with empty popup.

Changed code to scroll to top before popup display when content height of `ChoiceBox` is smaller than the scrolled offset.

Added system test to validate the fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8090123](https://bugs.openjdk.org/browse/JDK-8090123): Items are no longer visible when collection is changed


### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Committer)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/1039/head:pull/1039` \
`$ git checkout pull/1039`

Update a local copy of the PR: \
`$ git checkout pull/1039` \
`$ git pull https://git.openjdk.org/jfx pull/1039/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1039`

View PR using the GUI difftool: \
`$ git pr show -t 1039`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1039.diff">https://git.openjdk.org/jfx/pull/1039.diff</a>

</details>

lukostyra commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java:
```diff
@@ -810,6 +819,14 @@ Menu getOpenSubMenu() {
+    }
+
+    boolean isDownArrowVisible() {
+        return upArrow.isVisible();
```
Just to double-check, shouldn't this be `downArrow.isVisible()`?

karthikpandelu replied:
Yes it should be `downArrow.isVisible()`. I'll update the code.

karthikpandelu replied:
While this is a simple change, I thought of adding more checks on the visibility of both up and down arrow in the test to avoid these types of error in code. I'm seeing a strange issue that, I'm getting failures when more than 4 `Assert.assertTrue/assertFalse` statements are added. I'll investigate more on this and then update the code.

karthikpandelu replied:
Updated code and added more checks on visibility of up and down arrow in system test.

lukostyra commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Test
+    public void testChoicBoxScrollOnCollectionChange() throws Exception {
```
There's also a minor spelling error here (s/Choic/Choice)

karthikpandelu replied:
Fixed typo

lukostyra reviewed (commented):
I verified the changes on macOS and Windows - on both platforms `testChoicBoxScrollOnCollectionChange` passes when it's run as the only test (with `--tests '*testChoicBoxScrollOnCollectionChange'` option) but fails when you run all system tests:

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at test.robot.javafx.scene.ChoiceBoxScrollUpOnCollectionChangeTest.testChoicBoxScrollOnCollectionChange(ChoiceBoxScrollUpOnCollectionChangeTest.java:131)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	...
```

karthikpandelu commented:
> I verified the changes on macOS and Windows - on both platforms `testChoicBoxScrollOnCollectionChange` passes when it's run as the only test (with `--tests '*testChoicBoxScrollOnCollectionChange'` option) but fails when you run all system tests:
> 
> ```
> java.lang.AssertionError
> 	at org.junit.Assert.fail(Assert.java:87)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.junit.Assert.assertTrue(Assert.java:53)
> 	at test.robot.javafx.scene.ChoiceBoxScrollUpOnCollectionChangeTest.testChoicBoxScrollOnCollectionChange(ChoiceBoxScrollUpOnCollectionChangeTest.java:131)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
> 	...
> ```

I couldn't reproduce this issue both in windows and mac. Could you please give more details on this.

lukostyra reviewed (commented):
> I couldn't reproduce this issue both in windows and mac. Could you please give more details on this.

I rerun whole tests a few more times today on both platforms, each time with:

```
./gradlew clean
./gradlew sdk apps
./gradlew --continue --info -PFULL_TEST=true -PUSE_ROBOT=true test
```

On both platforms the results varied, sometimes it passed, others it failed. When it failed, it was assertions on `ChoiceBoxScrollUpOnCollectionChangeTest.java` @ line 131 or 132 (again, it wasn't consistently the same). I guess there is some sort of weird race condition going on.

As for environment details, on both systems I run tests on JDK 19.0.2. OS versions are macOS 12.6.1, Windows 11 22H2. Mac has default Retina XDR settings and scaling, Windows 11 machine has native resolution and 100% scaling. I made sure machines are awake and not turn the display off, screen save or go to sleep during tests. Also I made sure to not touch the mouse/trackpad on either one for the duration of the test.

I guess this would require checking it more in depth on someone else's machines. I'm not 100% sure what would cause this, since it's not consistent result it could be as well something in my environment that makes that specific test fail on some occasions.

Other than that, LGTM. However I think someone more experienced should take a closer look at it in order to approve it.

andy-goryachev-oracle reviewed (approved):
tested the fix with the MonkeyTester, this particular PR fixes the issue in the ticket.

I did notice two other issues with ChoiceBox (both present in the master as well as the fix, so both of them are unrelated to this PR):

1. adding ~1000 items takes a long time (1-2 seconds and a beach ball on Mac)
2. changing the data set does not clear the selected item, in fact it remains selected even though it may not be present in the new data set.

Do we have JBS bugs logged for these issues?  Karthik, could you please take a look?

this PR LG.

karthikpandelu commented:
> On both platforms the results varied, sometimes it passed, others it failed. When it failed, it was assertions on `ChoiceBoxScrollUpOnCollectionChangeTest.java` @ line 131 or 132 (again, it wasn't consistently the same). I guess there is some sort of weird race condition going on.
> 

Thanks for the details.
Other than UI components not getting loaded properly, i don't see other reasons for this inconsistent failure. I updated the test and added small delays to avoid these unexpected failures.
Added more number of items as well to the choicebox in the test so that while scrolling in larger screen we don't run out of options in the choicebox which results in hiding of down arrow which leads to test failure.

karthikpandelu commented:
> I did notice two other issues with ChoiceBox (both present in the master as well as the fix, so both of them are unrelated to this PR):
> 
>     1. adding ~1000 items takes a long time (1-2 seconds and a beach ball on Mac)
> 
>     2. changing the data set does not clear the selected item, in fact it remains selected even though it may not be present in the new data set.
> 
> 
> Do we have JBS bugs logged for these issues? Karthik, could you please take a look?

I couldn't find any bugs for these 2 issues in JBS. Created [JDK-8303060](https://bugs.openjdk.org/browse/JDK-8303060) and [JDK-8303064](https://bugs.openjdk.org/browse/JDK-8303064) to track these issues.

kevinrushforth commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java:
```diff
@@ -343,6 +343,15 @@ public void disposeContextMenu(ContextMenu menu) {
         itemsContainer.resize(w,contentHeight);
         itemsContainer.relocate(x, y);
 
+        if(contentHeight < Math.abs(ty)){
```
Minor: add space after `if`

kevinrushforth commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java:
```diff
@@ -810,6 +819,14 @@ Menu getOpenSubMenu() {
         return openSubmenu;
     }
 
+    boolean isUpArrowVisible() {
```
If these two new methods are only added for testing (which it looks like they are), please add a comment to that effect.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,183 @@
+                Toolkit.getToolkit().firePulse();
+            }
+
+            try {
+                Thread.sleep(400); // Wait for up arrow to get loaded in UI
+            } catch (Exception e) {}
```
Minor: If you use `Util.sleep` you don't need a try/catch.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,183 @@
+
+    private void scrollChoiceBox(int scrollAmt) throws Exception {
+        Util.runAndWait(() -> {
+            for (int i=0; i<scrollAmt; i++) {
```
Minor: add spaces around `=` and `<`.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,183 @@
+    private void addChoiceBoxItems(int count) {
+        Util.runAndWait(() -> {
+            ObservableList<String> items = FXCollections.observableArrayList();
+            for(int i=0 ; i<count ; i++) {
```
Minor: add space after `for` and spaces around `=` and `<`.

kevinrushforth reviewed (commented):
I added a few minor comments inline. I'll let others do the formal review.

karthikpandelu commented:
Addressed Kevin's comments.

andy-goryachev-oracle commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,181 @@
+
+    private void showChoiceBox() throws Exception {
+        mouseClick(choiceBox.getLayoutX() + choiceBox.getWidth() / 2,
+                    choiceBox.getLayoutY() + choiceBox.getHeight() / 2);
```
are these two lines slightly misaligned?

perhaps we could convert this to
double x = ...;
double y = ...;
mouseClick(x, y);

(there is a value in having one statement per line, in general)

karthikpandelu replied:
> are these two lines slightly misaligned?
> 
The second line is aligned at 6 tabs from the start of the line. Because of that it looks misaligned.

> perhaps we could convert this to double x = ...; double y = ...; mouseClick(x, y);
> 
> (there is a value in having one statement per line, in general)

Since there are no repeated function calls I didn't convert them to variables. Yes converting the long statements to variables would make it more readable. I will work on that.

karthikpandelu replied:
Updated code

andy-goryachev-oracle commented:
> Other than UI components not getting loaded properly, i don't see other reasons for this inconsistent failure. I updated the test and added small delays to avoid these unexpected failures.

I wonder if these failures are caused by JDK-8303060

karthikpandelu commented:
> I wonder if these failures are caused by JDK-8303060

It might be yes.

andy-goryachev-oracle commented:
Noticed a minor behavior issue, on Mac with multiple monitors.  The secondary monitor (scale=1) is positioned above the primary retina screen (scale=2).  When showing a popup in the case of 100 elements, the down arrow at the bottom cannot be seen.

In the attached screenshot, one can see the popup clipped at the secondary monitor bottom edge, the gray bar is the top of the primary screen:

![Screenshot 2023-02-23 at 09 44 11](https://user-images.githubusercontent.com/107069028/220988460-2939820c-108a-4000-9bf9-fc9145dfd68a.png)

andy-goryachev-oracle reviewed (approved):
The issue described in the ticket appears to be fixed.

karthikpandelu commented:
> Noticed a minor behavior issue, on Mac with multiple monitors. The secondary monitor (scale=1) is positioned above the primary retina screen (scale=2). When showing a popup in the case of 100 elements, the down arrow at the bottom cannot be seen.
> 

I couldn't reproduce this issue, but created bug [JDK-8303148](https://bugs.openjdk.org/browse/JDK-8303148) since you are seeing this issue.

kevinrushforth reviewed (commented):
The fix seems fine. The test is not completely stable, though. I ran it several times and it fails occasionally. I left a comment pointing to a problem in the test (sleeping on the wrong thread) that might explain it. Even if not, that needs to be fixed. I left a couple minor comments as well.

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,183 @@
+
+public class ChoiceBoxScrollUpOnCollectionChangeTest {
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+    static CountDownLatch scrollLatch = new CountDownLatch(1);
```
Used this latch now while stabilizing the test

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,183 @@
+                Toolkit.getToolkit().firePulse();
+            }
+
+            Util.sleep(400); // Wait for up arrow to get loaded in UI
```
Updated the code to sleep on the test thread.
Along with this, scrolling fixed amount (75) was also causing the occasional failures as the scroll event was getting triggered continuously and choice box was scrolling to the end making the down arrow getting hidden.
Made changes in code to scroll only required amount to make both up and down arrow visible based on the screen size and row height of choice box items.
Ran the test several times (around 15-20 times) in different screen sizes. Did not see failures now.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,191 @@
+ *
+ * There is 1 test in this file.
+ * Steps for testChoicBoxScrollOnCollectionChange()
+ * 1. Create a ChoiceBox and add 50 items to it.
```
Minor: the comment is wrong now (you add 150 items).

karthikpandelu replied:
Updated comment

kevinrushforth reviewed (commented):
The updated test seems stable now. I did leave a couple suggestions.

Have you run the test this on all platforms?

karthikpandelu commented:
> Have you run the test this on all platforms?

I have run the test in Mac and Windows multiple times. No failures found.

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,191 @@
+        Assert.assertTrue(ContextMenuContentShim.isContextMenuDownArrowVisible(popup));
+
+        double rowHeight = ContextMenuContentShim.getContextMenuRowHeight(popup);
+        double screenHeight = Screen.getPrimary().getBounds().getHeight();
```
Updated code to use `getVisualBounds()`

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/ChoiceBoxScrollUpOnCollectionChangeTest.java:
```diff
@@ -0,0 +1,191 @@
+
+        double rowHeight = ContextMenuContentShim.getContextMenuRowHeight(popup);
+        double screenHeight = Screen.getPrimary().getBounds().getHeight();
+        scrollChoiceBox((int) (screenHeight / rowHeight));
```
Updated code to use `Math.ceil()`

kevinrushforth replied:
The `Math.ceil` is in the wrong place. It's the result of the division that will generate the fractional result that needs it:

```
        double screenHeight = Screen.getPrimary().getVisualBounds().getHeight();
        scrollChoiceBox((int) Math.ceil(screenHeight / rowHeight));
```

karthikpandelu replied:
Updated code to use `Math.ceil` for the result of division. Please check.

andy-goryachev-oracle commented:
This PR looks good.  We can address [JDK-8303060](https://bugs.openjdk.org/browse/JDK-8303060) and [JDK-8303064](https://bugs.openjdk.org/browse/JDK-8303064) separately.

karthikpandelu commented:
/integrate

andy-goryachev-oracle commented:
/sponsor
